tower-http/src/validate_request.rs

```diff
@@ -0,0 +1,527 @@
+///
+/// See the [module docs](crate::validate_request) for an example.
+#[derive(Clone, Debug)]
+pub struct ValidateRequestHeader<S, T> {
+    inner: S,
+    validate: T,
+}
+
+impl<S, T> ValidateRequestHeader<S, T> {
+    fn new(inner: S, validate: T) -> Self {
+        Self::custom(inner, validate)
+    }
+
+    define_inner_service_accessors!();
+}
+
+impl<S, ResBody> ValidateRequestHeader<S, AcceptHeader<ResBody>> {
+    /// Validate requests have the required Accept header.
+    ///
+    /// The `Accept` header is required to be `*/*`, `type/*` or `type/subtype`,
+    /// as configured.
+    pub fn accept(inner: S, value: &str) -> Self
+    where
+        ResBody: Body + Default,
+    {
+        Self::custom(inner, AcceptHeader::new(value))
+    }
+}
+
+impl<S, T> ValidateRequestHeader<S, T> {
+    /// Validate requests using a custom method.
+    pub fn custom(inner: S, validate: T) -> ValidateRequestHeader<S, T> {
+        Self { inner, validate }
+    }
+}
+
+impl<ReqBody, ResBody, S, V> Service<Request<ReqBody>> for ValidateRequestHeader<S, V>
+where
+    V: ValidateRequest<ReqBody, ResponseBody = ResBody>,
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+{
+    type Response = Response<ResBody>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, ResBody>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        match self.validate.validate(&mut req) {
+            Ok(_) => ResponseFuture::future(self.inner.call(req)),
+            Err(res) => ResponseFuture::invalid_header_value(res),
+        }
+    }
+}
+
+pin_project! {
+    /// Response future for [`ValidateRequestHeader`].
+    pub struct ResponseFuture<F, B> {
+        #[pin]
+        kind: Kind<F, B>,
+    }
+}
+
+impl<F, B> ResponseFuture<F, B> {
+    fn future(future: F) -> Self {
+        Self {
+            kind: Kind::Future { future },
+        }
+    }
+
+    fn invalid_header_value(res: Response<B>) -> Self {
+        Self {
+            kind: Kind::Error {
+                response: Some(res),
+            },
+        }
+    }
+}
+
+pin_project! {
+    #[project = KindProj]
+    enum Kind<F, B> {
+        Future {
+            #[pin]
+            future: F,
+        },
+        Error {
+            response: Option<Response<B>>,
+        },
+    }
+}
+
+impl<F, B, E> Future for ResponseFuture<F, B>
+where
+    F: Future<Output = Result<Response<B>, E>>,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.project().kind.project() {
+            KindProj::Future { future } => future.poll(cx),
+            KindProj::Error { response } => {
+                /* Never panics unless polled after completion */
+                let response = response.take().unwrap();
```

Comment:
We usually write something like this:

```suggestion
                let response = response.take().expect("future polled after completion");
```

Its possible for users to poll futures after they've completed.